torchrl/modules/models/multiagent.py

```diff
@@ -158,22 +162,38 @@ def __init__(
         self.share_params = share_params
         self.centralised = centralised
 
-        self.agent_networks = nn.ModuleList(
-            [
-                MLP(
-                    in_features=n_agent_inputs
-                    if not centralised
-                    else n_agent_inputs * n_agents,
-                    out_features=n_agent_outputs,
-                    depth=depth,
-                    num_cells=num_cells,
-                    activation_class=activation_class,
-                    device=device,
-                    **kwargs,
-                )
-                for _ in range(self.n_agents if not self.share_params else 1)
+        def make_net():
+            return MLP(
+                in_features=n_agent_inputs
+                if not centralised
+                else n_agent_inputs * n_agents,
+                out_features=n_agent_outputs,
+                depth=depth,
+                num_cells=num_cells,
+                activation_class=activation_class,
+                device=device,
+                **kwargs,
+            )
+
+        if not self.share_params:
+            agent_networks = [
+                make_net()
+                for _ in range(self.n_agents)
             ]
-        )
+            self.params = TensorDictParams(torch.stack([TensorDict.from_module(mod) for mod in agent_networks], 0).contiguous(), no_convert=True)
+            net = agent_networks[0]
+        else:
+            net = make_net()
+            self.params = TensorDictParams(TensorDict.from_module(net), no_convert=True)
+        make_functional(net)
+        self.net = net
+        if self.centralised:
+            self.net_call = torch.vmap(
+                self.net,
+                (None, 0)
+            ) if not self.share_params else self.net
+        else:
+            self.net_call = torch.vmap(self.net, (-2, 0)) if not self.share_params else self.net
```

Comment:
Does this work and pass the test?

Cause it seems to me that we are not doing the resahping into (`n_agents` * `n_inputs_per_agent`) in the centralized case

Reply:
No it doesn't. In fact it's not ready for review :)

Reply:
But feel free to comment on the general idea from the PR description!